src/vsr/replica.zig

```diff
@@ -4899,112 +4948,322 @@ pub fn ReplicaType(
+
+            var v: ?u32 = null; // The highest `view_normal` from any replica.
+            var n: ?u64 = null; // The highest `op` for the highest `view_normal` from any replica.
+            var k: ?u64 = null; // The highest `commit_min` from any replica.
+
+            for (self.do_view_change_from_all_replicas) |received, replica| {
+                if (received) |message| {
+                    assert(message.header.command == .do_view_change);
+                    assert(message.header.cluster == self.cluster);
+                    assert(message.header.replica == replica);
+                    assert(message.header.view == self.view);
+
+                    // The view when this replica was last in normal status, which:
+                    // * may be higher than the view in any of the prepare headers.
+                    // * must be lower than the view of this view change.
+                    const view_normal = @intCast(u32, message.header.timestamp);
+                    assert(view_normal < message.header.view);
+
+                    if (replica == self.replica) {
+                        assert(view_normal == self.view_normal);
+                        assert(message.header.op == self.op);
+                        // We may have a newer commit than our own DVC due to async commits
+                        // (see below).
+                        assert(message.header.commit <= self.commit_min);
+                    }
+
+                    log.debug(
+                        "{}: on_do_view_change: " ++
+                            "replica={} view_normal={} op={} commit_min={}",
+                        .{
+                            self.replica,
+                            message.header.replica,
+                            view_normal,
+                            message.header.op,
+                            message.header.commit, // The `commit_min` of the replica.
+                        },
+                    );
+
+                    if (v == null or view_normal > v.?) {
+                        v = view_normal;
+                        n = message.header.op;
+                    } else if (view_normal == v.? and message.header.op > n.?) {
+                        n = message.header.op;
+                    }
+
+                    if (k == null or message.header.commit > k.?) k = message.header.commit;
+                }
+            }
+
+            // Consider the case:
+            // 1. Start committing op=N.
+            // 2. Send `do_view_change` to self.
+            // 3. Finish committing op=N.
+            // 4. Remaining `do_view_change` messages arrive, completing the quorum.
+            // In this scenario, the our own `do_view_change`'s commit is `N-1`, but `commit_min=N`.
+            // Don't let the commit backtrack.
+            if (k.? < self.commit_min) {
+                assert(k.? == self.do_view_change_from_all_replicas[self.replica].?.header.commit);
```

Comment:
Great!

Reply:
Ah, actually there was a bug in this assertion! It needs to handle committing multiple ops between while waiting for the DVC quorum.

Reply:
i.e. if we are replica 0, then replica 1's DVC commit_min might have been the highest in the quorum (`k`). But in the mean time we committed ahead of it (`self.commit_min > k`). I updated the comment, and the assert now reads:
```zig
                assert(self.commit_min >
                    self.do_view_change_from_all_replicas[self.replica].?.header.commit);
```

Reply:
Good catch. Expresses the concurrent commit_min increment better now too.